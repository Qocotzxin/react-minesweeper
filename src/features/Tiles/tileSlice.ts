import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Coordinates } from "../../types";

export interface TileState {
  availableFlags: number;
  flaggedTiles: Coordinates[];
}

const initialState: TileState = {
  availableFlags: 10,
  flaggedTiles: [],
};

export const tileSlice = createSlice({
  name: "tile",
  initialState,
  reducers: {
    updateFlaggedTiles(state, action: PayloadAction<Coordinates>) {
      const exists = state.flaggedTiles.find((tile) => {
        return tile.x === action.payload.x && tile.y === action.payload.y;
      });
      if (!exists) {
        if (state.availableFlags <= 0) {
          return;
        }
        state.flaggedTiles.push(action.payload);
        state.availableFlags -= 1;
      } else {
        const updateFlaggedTiles = state.flaggedTiles.findIndex(
          (tile) => tile.x === action.payload.x && tile.y === action.payload.y
        );
        state.flaggedTiles.splice(updateFlaggedTiles, 1);
        state.availableFlags += 1;
      }
    },
    selectCoordinates(state, action: PayloadAction<Coordinates>) {},
    resetFlags(state) {
      state.availableFlags = 10;
      state.flaggedTiles = [];
    },
  },
});

export const { selectCoordinates, updateFlaggedTiles, resetFlags } =
  tileSlice.actions;

export default tileSlice.reducer;
