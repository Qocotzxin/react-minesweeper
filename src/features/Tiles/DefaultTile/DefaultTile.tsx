import { Box, CircularProgress } from "@mui/material";
import { FC, KeyboardEvent, SyntheticEvent, useState } from "react";
import { HiFlag } from "react-icons/hi";
import { useSelector } from "react-redux";
import { useAppDispatch } from "../../../app/hooks";
import { RootState } from "../../../app/store";
import { selectCoordinates, updateFlaggedTiles } from "../tileSlice";

interface DefaultTileProps {
  colIndex: number;
  rowIndex: number;
  disabled?: boolean;
}

export const DefaultTile: FC<DefaultTileProps> = ({
  colIndex,
  rowIndex,
  disabled,
}) => {
  const [isLoading, setIsLoading] = useState(false);
  const dispatch = useAppDispatch();
  const flaggedTiles = useSelector(
    (state: RootState) => state.tile.flaggedTiles
  );

  const isFlagged = flaggedTiles.find(
    (tile) => tile.x === colIndex && tile.y === rowIndex
  );

  /**
   * Executes the click prop callback passing the coordinates.
   */
  const onTileClick = () => {
    setIsLoading(true);
    dispatch(selectCoordinates({ x: colIndex, y: rowIndex }));
    if (isFlagged) {
      dispatch(updateFlaggedTiles({ x: colIndex, y: rowIndex }));
    }
  };

  /**
   * Adds a flag to the block when right clicking.
   */
  const onContextMenu = (event: SyntheticEvent) => {
    event.preventDefault();
    dispatch(updateFlaggedTiles({ x: colIndex, y: rowIndex }));
  };

  const onKeyUp = (event: KeyboardEvent<HTMLDivElement>) => {
    if (event.key === "Enter") {
      onTileClick();
    }
  };

  return (
    <Box
      aria-label={`Tile ${colIndex} ${rowIndex}`}
      onKeyUp={onKeyUp}
      tabIndex={0}
      data-testid="DefaultTile"
      bgcolor="rgb(148, 151, 156)"
      border="solid #808080"
      width="1.5rem"
      height="1.5rem"
      display="flex"
      justifyContent="center"
      alignItems="center"
      onClick={onTileClick}
      onContextMenu={onContextMenu}
      sx={{
        position: "relative",
        borderWidth: " 0 1px 1px 0",
        boxShadow: 5,
        cursor: "pointer",
        pointerEvents: disabled ? "none" : "initial",
        "&:hover": {
          background: "#d0d0d0",
        },
      }}
    >
      {isFlagged && !isLoading && (
        <HiFlag
          size={18}
          color="rgb(201, 91, 118)"
          data-testid="DefaultTile-flag"
        />
      )}
      {isLoading && <CircularProgress color="secondary" size={18} />}
    </Box>
  );
};
