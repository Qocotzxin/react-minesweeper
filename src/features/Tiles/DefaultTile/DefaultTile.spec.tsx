import { fireEvent, render, screen } from "@testing-library/react";
import { DefaultTile } from "./DefaultTile";
import { configureStore } from "@reduxjs/toolkit";
import * as tiles from "../tileSlice";
import { Provider } from "react-redux";
import { FC } from "react";
import tileReducer from "../tileSlice";

const DefaultTileMockProps = {
  colIndex: 1,
  rowIndex: 2,
  disabled: false,
};

afterAll(() => {
  jest.resetAllMocks();
});

interface WrapperProps {
  preloadedState?: tiles.TileState;
  disabled?: boolean;
}

const defaultState = {
  availableFlags: 10,
  flaggedTiles: [],
};

const Wrapper: FC<WrapperProps> = ({
  preloadedState = defaultState,
  disabled,
}) => {
  return (
    <Provider
      store={configureStore({
        reducer: { tile: tileReducer },
        preloadedState: { tile: preloadedState },
      })}
    >
      <DefaultTile {...DefaultTileMockProps} disabled={disabled} />
    </Provider>
  );
};

describe("DefaultTile", () => {
  describe("Snapshot", () => {
    it("Should render correctly and match snapshot.", () => {
      const { container } = render(<Wrapper />);

      expect(container).toMatchSnapshot();
    });
  });

  describe("Behavior", () => {
    it("Should call selectCoordinates with the right coordinates when tile is clicked.", () => {
      render(<Wrapper />);
      const spy = jest.spyOn(tiles, "selectCoordinates");

      fireEvent.click(screen.getByTestId("DefaultTile"));

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith({ x: 1, y: 2 });
    });

    it("Should call updateFlaggedTiles with the right coordinates when tile is right clicked.", () => {
      render(<Wrapper />);
      const spy = jest.spyOn(tiles, "updateFlaggedTiles");

      fireEvent.contextMenu(screen.getByTestId("DefaultTile"));

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith({ x: 1, y: 2 });
    });

    it("Should display a spinner when the button is clicked.", () => {
      render(<Wrapper />);

      fireEvent.click(screen.getByTestId("DefaultTile"));

      expect(screen.getByRole("progressbar")).toBeInTheDocument();
    });

    it("Should set pointerEvents to none when the button is disabled.", () => {
      render(<Wrapper disabled />);

      expect(screen.getByTestId("DefaultTile")).toHaveStyle({
        pointerEvents: "none",
      });
    });

    it("Should add a flag if available when right clicking on element.", () => {
      render(<Wrapper disabled />);

      fireEvent.contextMenu(screen.getByTestId("DefaultTile"));

      expect(screen.getByTestId("DefaultTile-flag")).toBeInTheDocument();
    });
  });
});
