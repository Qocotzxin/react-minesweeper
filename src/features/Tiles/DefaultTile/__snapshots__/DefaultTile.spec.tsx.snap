// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultTile Snapshot Should render correctly and match snapshot. 1`] = `
<div>
  <div
    aria-label="Tile 1 2"
    class="MuiBox-root css-7wl9ds"
    data-testid="DefaultTile"
    tabindex="0"
  />
</div>
`;
